### ISSUE TITLE
row_major matrix has incorrect printing

### ISSUE THREAD

**Issue by KalinMagus** [odin-lang/Odin#4210]
## Context

Printing a non-square row_major matrix results in incorrect information. Not sure if this is just `fmt` or a deeper issue.
```
 	Odin:    dev-2024-09:90358f97d
	OS:      Arch Linux, Linux 6.10.6-arch1-1
	CPU:     AMD Ryzen 9 5900X 12-Core Processor            
	RAM:     64243 MiB
	Backend: LLVM 18.1.8
```
### Code to Reproduce
```go
package test

import "core:fmt"

main :: proc()
{
	row_major34: #row_major matrix[3,4]int = {
		11,12,13,14,
		21,22,23,24,
		31,32,33,34,
	}
	row_major43: #row_major matrix[4,3]int = {
		11,12,13,
		21,22,23,
		31,32,33,
		41,42,43,
	}

	fmt.printfln("row_major34 = %#v", row_major34)
	fmt.printfln("row_major43 = %#v", row_major43)
}
```
The last values of row_major34 seem random. I suspect they are just unused stack space.
#### Expected Output
```
row_major34 = matrix[
	11, 12, 13, 14,
	21, 22, 23, 24,
	31, 32, 33, 34,
]
row_major43 = matrix[
	11, 12, 13,
	21, 22, 23,
	31, 32, 33,
	41, 42, 43,
]
```
#### Current Output
```
row_major34 = matrix[
	11, 12, 13, 21,
	21, 22, 23, 31,
	31, 32, 33, 34,
]
row_major43 = matrix[
	11, 12, 13,
	21, 22, 23,
	31, 32, 33,
	3573412791104, 3573412791104, 3573412791104,
]
```

**Comment by Kelimion** [comment]
I've confirmed that `row_major43` (not `row_major34` as mistakenly stated) prints random garbage.

**Comment by KalinMagus** [comment]
Sorry, I should have been clearer. `row_major34` isn't random like `row_major43`, but is still incorrect. Specifically, the last column seems to use the value from the first column for the first two rows, but the fourth row's last value is correct for some reason.

**Comment by Kelimion** [comment]
I see what you mean. I'm seeing that as well.

**Comment by Kelimion** [comment]
The type info's stride looks to be correct:
```cpp
ti(row_major34): Type_Info_Matrix{
	elem = int,
	elem_size = 8,
	elem_stride = 3,
	row_count = 3,
	column_count = 4,
	layout = "Column_Major",
}
ti(row_major43): Type_Info_Matrix{
	elem = int,
	elem_size = 8,
	elem_stride = 4,
	row_count = 4,
	column_count = 3,
	layout = "Column_Major",
}
```

**Comment by KalinMagus** [comment]
On a further look it appears the data might be wrong not just the printing. Adding the following code to main seems to indicate the data in the matrix is actually wrong. Though not in a way consistent with the printing. Not sure how often `#row_major` is used so it might just be wrong in the compiler itself
```
	ptr34 := transmute(^[3*4]int)&row_major34
	ptr43 := transmute(^[4*3]int)&row_major43
	fmt.printfln("ptr34 = %#v", ptr34)
	fmt.printfln("ptr43 = %#v", ptr43)
```

**Comment by Kelimion** [comment]
I just came to that conclusion as well:
```cpp
package repro

import "core:fmt"
@(require) import "base:intrinsics"
@(require) import "base:runtime"

main :: proc() {
	row_major34: #row_major matrix[3,4]int = {
		11,12,13,14,
		21,22,23,24,
		31,32,33,34,
	}
	row_major43: #row_major matrix[4,3]int = {
		11,12,13,
		21,22,23,
		31,32,33,
		41,42,43,
	}
	fmt.printfln("row_major34 = %#v\nflat: %v", row_major34, intrinsics.matrix_flatten(row_major34))
	fmt.printfln("row_major43 = %#v\nflat: %v", row_major43, intrinsics.matrix_flatten(row_major43))
```
```
row_major34 = matrix[
	11, 12, 13, 21,
	21, 22, 23, 31,
	31, 32, 33, 34,
]
flat: [11, 21, 31, 12, 22, 32, 13, 23, 33, 21, 31, 34]
row_major43 = matrix[
	11, 12, 13,
	21, 22, 23,
	31, 32, 33,
	28992425118793820, 28992425118793820, 28429488050470958,
]
flat: [11, 21, 31, 28992425118793820, 12, 22, 32, 28992425118793820, 13, 23, 33, 28429488050470958]
```

**Comment by Kelimion** [comment]
On the plus side it's easily tested for in the CI by constructing a couple of minor and major matrices and using `matrix_flatten`, checking if we get out the expected numbers.

**Comment by KalinMagus** [comment]
I will point out that `matrix_flatten` and using `transmute(^[4*3]int)` gives me *different* values. So there might be multiple bugs here.

**Comment by Kelimion** [comment]
Bill fixed it in 0a08a65202ca40857e580036c561bcbf940a6870 and I just added CI tests exercising both `(^[M*N]E)` and `matrix_flatten` to get the values.
